### ISSUE TITLE
Problem when applying a Phong shading on my TrueType font

### ISSUE THREAD

**Issue by rchastain** [bgrabitmap/bgrabitmap#84]
Hello Johann!

It's probably not a BGRABitmap issue, but I would need your help to understand something.
I generate PNG pictures, with a Phong shading, from TrueType fonts. I have a problem with one font (Montreal). What is weird is that, without Phong shading, there is no problem.
If you have time, could you please take a look at it?

- [Pictures without shading](https://drive.google.com/open?id=1Mqj2XA7zIi0DPeCAauv2xuBC386E-vPF)
- [Pictures with shading](https://drive.google.com/open?id=1USzoKN9ESmIUvAY3ipXKWBbHj88B2UTj)

If you look at pictures in the _montreal_ folder, you will see what I mean.
Both packages include source code. The second package also includes TTF files.

Regards.

Roland

**Comment by rchastain** [comment]
With some pictures attached it will be clearer. :)

![bb](https://user-images.githubusercontent.com/13957204/72215622-66975300-3515-11ea-9d7d-bfbcde0afa1d.png)
![bb](https://user-images.githubusercontent.com/13957204/72215618-5a12fa80-3515-11ea-8017-483e772e7e2d.png)

**Comment by circular17** [comment]
Indeed, it seems there is a problem with the size of the font. Is the code browsable on Git?

**Comment by rchastain** [comment]
Thank you for your answer. The source code of the Lazarus project is included in the ZIP archive.

**Comment by circular17** [comment]
Ok. Where in the code do you prepare the images?

**Comment by rchastain** [comment]
Here?
```
  LContour := TBGRABitmap.Create(CSize, CSize, BGRAPixelTransparent);
  LShader := TPhongShading.Create;
  LRenderer := TBGRAFreeTypeFontRenderer.Create(LShader, True);
  LRenderer.ShadowVisible := True;
  LContour.FontRenderer := LRenderer;
  LContour.FontName := CFonts[LFont].name;
  LContour.FontFullHeight := CSize;
  LContour.FontQuality := fqFineClearTypeRGB;
```

**Comment by rchastain** [comment]
OK, I found.

```
  //LContour.FontFullHeight := CSize;
  LContour.FontHeight := CSize;
```
But I wonder why the problem didn't happen with other fonts.

What is the difference between _FontFullHeight_ and _FontHeight_?

**Comment by rchastain** [comment]
No, there is still a problem (and still no problem with other fonts).
![bk](https://user-images.githubusercontent.com/13957204/72217654-33fb5380-3531-11ea-859e-32b08b80e7f3.png)

**Comment by circular17** [comment]
FontHeight or FontEmHeight are the size of the letters themselves (it used to be the width of the letter M but not always anymore). That's the size the is shown in most text editors.

FontFullHeight is the total height the text line, including all vertical spacing.

For a given font, it is more reliable to set the FontEmHeight because the line spacing may vary depending on the font engine. Though such as size will not give the same letter size among different fonts. Conversely if you're interesting in fitting the text vertically, FontFullHeight will be more reliable, though the spacing may be too large in some cases.

Can you tell me where in the files I can find both the shaded version and the non shaded version, so that I can compare?

**Comment by rchastain** [comment]
Thank you for the explanation.
_chessimage1.zip_ is the non shaded version. _chessimage2.zip_ is the shaded version.

**Comment by circular17** [comment]
Yeah but more specifically, what file and what line?

**Comment by rchastain** [comment]
For the project _chessimage1_, please see _chessimage.lpr_, line 71:
```
    contour := TBGRABitmap.Create(CSize, CSize, BGRAPixelTransparent);
    contour.FontName := CFonts[font].name;
    contour.FontHeight := CSize;
    contour.TextOut(0, 0, CTable[i, CFonts[font].index], BGRABlack);
```

For the project _chessimage2_, please see _chessimage_u.pas_, line 124:
```
  LContour := TBGRABitmap.Create(CSize, CSize, BGRAPixelTransparent);
  LShader := TPhongShading.Create;
  LRenderer := TBGRAFreeTypeFontRenderer.Create(LShader, True);
  LRenderer.ShadowVisible := True;
  LContour.FontRenderer := LRenderer;
  LContour.FontName := CFonts[LFont].name;
  //LContour.FontFullHeight := CSize;
  LContour.FontHeight := CSize;
  LContour.FontQuality := fqFineClearTypeRGB;
```

**Comment by circular17** [comment]
Ok so in theory that should be the same.

**Comment by rchastain** [comment]
> Ok so in theory that should be the same.

Yes, that's why I don't understand. But I don't want to bore you with that. There is no emergency. I was curious to understand what happens.

Regards.

Roland

**Comment by circular17** [comment]
I will look more closely when I can.

Regards

**Comment by rchastain** [comment]
Thank you.

**Comment by circular17** [comment]
Hello, it seems that Montreal font has not a well-defined size. When I open it in a font viewer it goes outside of the line. 

Though I am not sure why this would be handled differently. 

Regarding the FontFullHeight / FontEmHeight, in the cases of chess font, it seems that they use the whole line, so that it does not really matter which one to use. Though I would suggest to use the same anyway in all cases.

I cannot compile the chessimage2 so I cannot debug the program. I get a linking problem.

I noticed that in one case you use BGRABitmapPack4NoGui and not in the other. I suggest you try to incorporate the effect in the chessimage1 (by setting a boolean for example) so that there is minimal difference when comparing.

Regards

**Comment by rchastain** [comment]
Hello! Thank you for your answer. OK, I will say what you suggest. I come back as soon as it made.

Regards.

Roland

**Comment by rchastain** [comment]
Here is [chessimage2 modified](https://drive.google.com/open?id=1eceF_aFKSvXL7ouXCTMWsW-dLsSSj28W) (console application, BGRABitmapPack4NoGUI).
I didn't include the _fonts_ directory. Please use the fonts from the previous package.

As you will see, the problem is still here.

Thank you for your kind assistance.

Regards.

Roland

**Comment by circular17** [comment]
Ok, so. In fact, the original problem was due to the font full height difference. Because the glyphs are oversized relative to the line size, using the full height will make the letters bigger. Thus it overflows on both right and bottom.

Using the em height solves the size problem. But now there is another problem which is that the rendering is clipped. The glyph is rendered up to the line limit. The reason is that the text effect is computed in a temporary bitmap which has the size of the text line.

To sum up, the font is not well defined, and the limitations of the text engine gives this result. One might improve that a bit, by using the actual glyph size to determine the size of the temporary bitmap and at the same time restrict that to the clipping rectangle of the destination bitmap.

That may be a bit of work, so I am not going to do that now. Maybe in the future. In particular, this might be necessary when using very big size of letters. As of now, using huge letters will create a huge temporary bitmap even though only a small portion will be visible in the destination bitmap.

**Comment by rchastain** [comment]
Thank you. All is clear now.

I don't know if I should close the issue or not, so I let you do. See you soon!

Regards.

Roland

**Comment by rchastain** [comment]
If you don't mind, I will remove the ZIP files from my Google Drive and delete the links in the discussion, because Chess Montreal isn't supposed to be freely distributed. So please keep a copy of it if you want to experiment.
Regards.
Roland

**Comment by circular17** [comment]
Ok no problemo.
